### ISSUE TITLE
Identifying assets

### ISSUE THREAD

**Issue by LourensVeen** [SecConNet/mahiru#18]
In this system, applications and data are represented by assets. An asset will be implemented as a container image. Assets are added to the system by one of the participants (parties), by putting the asset into their AssetStore. Users can use an asset by referring to it in a workflow they submit. During planning of the workflow, permissions for the assets involved are looked up in a PolicyStore. During execution of the workflow, the asset will be retrieved from the AssetStore by a Runner, and instantiated into a container. Parties may cache copies of frequently used assets from elsewhere in their own AssetStore, so that they don't have to re-download them all the time.

Assets need to be identified somehow. We can distinguish here between primary assets, which come from outside the system and are put into it, and secondary assets, which have been created as a (final or intermediate) result of a workflow. Currently, primary assets are identified by an identifier of the form `id:<namespace>/<path>`. Both namespaces and paths may contain slashes, so splitting them requires a list of registered namespaces. This inconvenient, because we need to extract the namespace to e.g. figure out who owns the asset (it's the owner of the namespace).

For secondary assets, we currently use an id of the form `hash:<hash>`, where the hash is a hexadecimal string. The hash of a step output is generated by combining the hashes of the inputs and the compute asset, and hashing that in turn. As a result, if two submitted workflows produce an (intermediate) output in the same way, then those outputs will be identical (absent non-deterministic compute assets!) and will have the same hash. Also, this means that for any given workflow, the hashes for the (intermediate) results can be calculated from the workflow description itself. Thus, if subsequent steps are executed in different locations, then all we need is to send the whole workflow to both locations with a description of where each step is executed, and then they can each calculate the ids of the intermediate assets that they need to publish and retrieve, and find out where to get them.

This works, but there is one issue, which is that workflows have inputs which are primary assets. So a job consists of the workflow and the ids of the primary assets to feed into it. Unfortunately, we cannot determine from that id in which AssetStore it is stored, and so we cannot retrieve it.

Currently, this is worked around by registering the assets and their location in the global registry, but this doesn't scale. It is also undesirable from the perspective of a secure federated digital data marketplace, because it means that all assets are globally visible, while participants may want to share a particular asset with specific other participants, but not disclose its existence to anyone else.

There are several ways to resolve this. One would be to just ask all known AssetStores for the required asset. That's linear in the number of participants, which is not so nice, and it causes unnecessary extra load on the AssetStores. Also, it advertises the existence of a particular asset to other participants than the one executing the step (which has access permissions anyway) and the owner of the asset. Since the current ids are somewhat meaningful, this would be an information leak.

So what if we put the AssetStore in which the asset is primarily stored in the identifier itself? That would make it an [intrinsic identifier](https://www.softwareheritage.org/2020/07/09/intrinsic-vs-extrinsic-identifiers/). We also need the namespace in plain text, because we need to be able to determine the PolicyServer to access to get rules for this asset from, and that's looked up by the namespace it serves rules for. For the path within the namespace, we could actually blind it by hashing, or actually even generate a random identifier (or even better, use a hash of the image!), and store the name separately as metadata. The latter can be done with the current system, and it leaves the choice with the asset owner, which is nice.

What if an asset is copied to another system and stored there in an AssetStore? Then its id won't match its location. But that's fine, you can still look it up in the canonical store, and that's probably good anyway. And you could also still search other AssetStores for whether they have a copy. At least you always have a place that's guaranteed to have it.

So, primary assets will have an idea of the form `id:<AssetStoreId>:<namespace>:<name>`.

- [ ] Implement this
- [ ] Remove asset registration

**Comment by LourensVeen** [comment]
Actually, a colon is probably not a good separator, as it is a reserved character in URIs, and we have REST APIs where one may want to access `assets/{id}`. Of course we can urlencode, but it would be nicer if ids could be used directly. This does limit the choice rather a lot though: only hypen, period, underscore and tilde are always okay to put into a URI.

So we could have either of these:
```
id-<AssetStoreId>-namespace-name
id.<AssetStoreId>.namespace.name
id_AssetStoreId_namespace_name
id~AssetStoreId~namespace~name
```
If we want to use domain names for namespaces then those will have periods in them, so best not use those. Underscores may be used to separate words inside a name, same with dashes. Those tildes scare me.

So maybe we should just make sure to urlencode them when used in a URL?

**Comment by LourensVeen** [comment]
There's also the issue about which characters are allowed in between the colons. Slashes seem to be out, as these are taken as path separators even when encoded (see https://github.com/falconry/falcon/issues/897). Maybe letters, underscores, dashes and periods will be good enough?

**Comment by LourensVeen** [comment]
We've now changed to registering whole sites, so instead of an asset store id, we need a site name. So we get `id:<site>:<namespace>:<name>`. And I think I like having the site at the end better. So

`id:<namespace>:<name>:<site>`

it is.

**Comment by LourensVeen** [comment]
Actually, there's something missing here. We also have asset collections, which are assets but abstract ones, in that they cannot be downloaded. They're primary assets though, not secondary. So if we put the site at the end and make it optional, then we can have

`id:<namespace>:<name>:<site>`

for concrete assets and

`id:<namespace>:<name>`

for abstract assets.

**Comment by LourensVeen** [comment]
Closed by #46.
